include/deal.II/lac/trilinos_tpetra_block_sparse_matrix.templates.h

```diff
@@ -0,0 +1,340 @@
+
+      // ... set the correct sizes, ...
+      this->row_block_indices    = block_sparsity_pattern.get_row_indices();
+      this->column_block_indices = block_sparsity_pattern.get_column_indices();
+
+      // ... and then assign the correct
+      // data to the blocks.
+      for (size_type r = 0; r < this->n_block_rows(); ++r)
+        for (size_type c = 0; c < this->n_block_cols(); ++c)
+          {
+            this->sub_objects[r][c]->reinit(parallel_partitioning[r],
+                                            parallel_partitioning[c],
+                                            block_sparsity_pattern.block(r, c),
+                                            communicator,
+                                            exchange_data);
+          }
+    }
+
+
+
+    template <typename Number, typename MemorySpace>
+    template <typename BlockSparsityPatternType>
+    void
+    BlockSparseMatrix<Number, MemorySpace>::reinit(
+      const BlockSparsityPatternType &block_sparsity_pattern)
+    {
+      std::vector<IndexSet> parallel_partitioning;
+      for (size_type i = 0; i < block_sparsity_pattern.n_block_rows(); ++i)
+        parallel_partitioning.emplace_back(
+          complete_index_set(block_sparsity_pattern.block(i, 0).n_rows()));
+
+      reinit(parallel_partitioning, block_sparsity_pattern);
+    }
+
+
+
+    template <typename Number, typename MemorySpace>
+    void
+    BlockSparseMatrix<Number, MemorySpace>::reinit(
+      const std::vector<IndexSet>               &parallel_partitioning,
+      const ::dealii::BlockSparseMatrix<double> &dealii_block_sparse_matrix,
+      const MPI_Comm                             communicator,
+      const double                               drop_tolerance)
+    {
+      const size_type n_block_rows = parallel_partitioning.size();
+
+      Assert(n_block_rows == dealii_block_sparse_matrix.n_block_rows(),
+             ExcDimensionMismatch(n_block_rows,
+                                  dealii_block_sparse_matrix.n_block_rows()));
+      Assert(n_block_rows == dealii_block_sparse_matrix.n_block_cols(),
+             ExcDimensionMismatch(n_block_rows,
+                                  dealii_block_sparse_matrix.n_block_cols()));
+
+      // Call the other basic reinit function ...
+      reinit(n_block_rows, n_block_rows);
+
+      // ... and then assign the correct
+      // data to the blocks.
+      for (size_type r = 0; r < this->n_block_rows(); ++r)
+        for (size_type c = 0; c < this->n_block_cols(); ++c)
+          {
+            this->sub_objects[r][c]->reinit(parallel_partitioning[r],
+                                            parallel_partitioning[c],
+                                            dealii_block_sparse_matrix.block(r,
+                                                                             c),
+                                            communicator,
+                                            drop_tolerance);
```

Comment:
I think it is this place. The function takes a block sparse matrix as argument, and so here we call `reinit()` with a sparse matrix as third argument, but the function expects a sparsity pattern.

I think what is necessary is to provide an overload of `LA::TpetraWrappers::SparseMatrix::reinit (..., dealii::SparseMatrix, ...).